pkg/gauges/contracts/gauges/PremintedGauge.sol

```diff
@@ -0,0 +1,137 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/IERC20.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/ReentrancyGuard.sol";
+
+import "../interfaces/IBalancerMinter.sol";
+import "../interfaces/IBalancerTokenAdmin.sol";
+import "../interfaces/IGaugeController.sol";
+import "../interfaces/ILiquidityGauge.sol";
+
+abstract contract PremintedGauge is ILiquidityGauge, ReentrancyGuard {
+    IERC20 internal immutable _balToken;
+    IBalancerTokenAdmin private immutable _tokenAdmin;
+    IBalancerMinter private immutable _minter;
+    IGaugeController private immutable _gaugeController;
+
+    uint256 private immutable _INITIAL_RATE;
+    uint256 private immutable _RATE_REDUCTION_TIME;
+    uint256 private immutable _RATE_REDUCTION_COEFFICIENT;
+    uint256 private immutable _RATE_DENOMINATOR;
+
+    uint256 private _rate;
+    uint256 private _period;
+    uint256 private _startEpochTime;
+
+    uint256 private _emissions;
+    bool public isKilled;
+
+    constructor(IBalancerMinter minter) {
+        IBalancerTokenAdmin tokenAdmin = IBalancerTokenAdmin(minter.getBalancerTokenAdmin());
+        IERC20 balToken = tokenAdmin.getBalancerToken();
+
+        _balToken = balToken;
+        _tokenAdmin = tokenAdmin;
+        _minter = minter;
+        _gaugeController = minter.getGaugeController();
+
+        _INITIAL_RATE = tokenAdmin.INITIAL_RATE();
+        _RATE_REDUCTION_TIME = tokenAdmin.RATE_REDUCTION_TIME();
+        _RATE_REDUCTION_COEFFICIENT = tokenAdmin.RATE_REDUCTION_COEFFICIENT();
+        _RATE_DENOMINATOR = tokenAdmin.RATE_DENOMINATOR();
+
+        // Because we calculate the rate locally, this gauge cannot
+        // be used prior to the start of the first emission period
+        uint256 rate = tokenAdmin.rate();
+        require(rate != 0);
+
+        _rate = rate;
+        _period = _currentPeriod();
+        _startEpochTime = tokenAdmin.startEpochTimeWrite();
+    }
+
+    function checkpoint() external nonReentrant returns (bool) {
+        // TODO: Add a guard to checkpointing.
+        uint256 lastPeriod = _period;
+        uint256 currentPeriod = _currentPeriod();
+
+        if (lastPeriod < currentPeriod) {
+            _gaugeController.checkpoint_gauge(address(this));
+
+            uint256 rate = _rate;
+            uint256 newEmissions = 0;
+            lastPeriod += 1;
+            uint256 nextEpochTime = _startEpochTime + _RATE_REDUCTION_TIME;
+            for (uint256 i = lastPeriod; i < lastPeriod + 255; ++i) {
+                if (i > currentPeriod) break;
+
+                uint256 periodTime = i * 1 weeks;
+                uint256 periodEmission = 0;
+                uint256 gaugeWeight = _gaugeController.gauge_relative_weight(address(this), periodTime);
+
+                if (nextEpochTime >= periodTime && nextEpochTime < periodTime + 1 weeks) {
```

Comment:
```suggestion
                if (nextEpochTime >= periodTime && nextEpochTime < (periodTime + 1 weeks)) {
```

Dependency on operator precedence makes me very anxious.

Reply:
This is handled by the type system however, `true + 1 weeks` would throw an error.

Reply:
C scarred me for life